theano/tensor/basic.py

```diff
@@ -4417,17 +4417,12 @@ def infer_shape(self, node, ishapes):
                                  ' must have at most one entry equal to -1')
             return [requ]
         else:
-            oshape = []
-            for i in xrange(self.ndim):
-                default_os_i = theano.tensor.opt.Shape_i(i)(node.outputs[0])
-                try:
-                    os_i = get_scalar_constant_value(node.inputs[1][i]).item()
-                    if os_i == -1:
-                        os_i = default_os_i
-                except NotScalarConstantError:
-                    os_i = default_os_i
-                oshape.append(os_i)
-            return [tuple(oshape)]
+            new_dims = [node.inputs[1][i] for i in xrange(self.ndim)]
+            return [tuple([switch(eq(new_dims[i], -1),
+                                  theano.tensor.mul(*ishapes[0]) /
+                                  (-theano.tensor.mul(*new_dims)),
```

Comment:
At least some failures in travis are coming from the fact that one of these `mul(*...)` is called with no arguments, meaning that `new_dims` or `ishapes[0]` is empty.
A way to fix that would be to replace those with 1.